t/js_saml.t

```diff
@@ -0,0 +1,1642 @@
+        }
+    } else {
+        die "Unknown SAML message type: $type";
+    }
+
+    return $xml_obj;
+}
+
+sub send_saml {
+    my ($xml_obj, $dst, %extra) = @_;
+    my ($r, $b64);
+
+    $dst //= $acs;
+
+    $XML::LibXML::skipXMLDeclaration = 1;
+    my $xml_str = $xml_obj->toString();
+
+    if (exists($extra{method}) && $extra{method} eq 'get') {
+        my $compressed_xml;
+        rawdeflate(\$xml_str => \$compressed_xml)
+            or die "rawdeflate failed: $RawDeflateError\n";
+        $b64 = encode_base64($compressed_xml, '');
+        my $url = $dst . "?SAMLResponse=" . uri_escape($b64) .
+            ($extra{relay_state} ? "&RelayState=" . $extra{relay_state} : "");
+        $r = get($url, %extra);
+    } else {
+        $b64 = encode_base64($xml_str, '');
+        my $body = "SAMLResponse=" . uri_escape($b64) .
+            ($extra{relay_state} ? "&RelayState=" . $extra{relay_state} : "");
+        $r = http_post($dst, $body, %extra);
+    }
+
+    return $r;
+}
+
+sub cfg_post {
+    my ($arg, $post, $host) = @_;
+    $host //= 'sp.example.com';
+    my $json;
+
+    if (ref $arg eq 'HASH') {
+        $json = $arg;
+    } elsif (!ref $arg) {
+        my ($key, $value) = each %{decode_json($arg)};
+        $cfg->{$key} = $value;
+        $json = {$key => $value};
+    } else {
+        die "Invalid arguments for cfg_post";
+    }
+
+    for my $key (keys %$json) {
+        my $value = $json->{$key};
+        my $data = { $host => $value };
+        my $data_string = encode_json($data);
+
+        if ($post) {
+            http_post("$kv/$key", $data_string);
+        } else {
+            http_patch("$kv/$key", $data_string);
+        }
+    }
+}
+
+sub cfg_verify {
+    my ($param, $test_name) = @_;
+
+    my $url = "$kv/$param";
+    my $original_value = getkv($url);
+
+    http_patch($url, '{"sp.example.com": ""}');
+    my $r = get('/login');
+    http_patch($url, $original_value);
+
+    like($r, qr/(?=.*Invalid)(?=.*$param)/s, $test_name);
+
+    return $r;
+}
+
+sub http_post {
+	my ($url, $body, %extra) = @_;
+	my $len = length($body);
+
+    my $auth_token = $extra{auth_token} || '';
+    my $auth_redir = $extra{auth_redir} || '';
+
+	http(<<EOF);
+POST $url HTTP/1.0
+Host: sp.example.com
+Content-Length: $len
+Cookie: auth_token=$auth_token
+Cookie: auth_redir=$auth_redir
+Content-Type: application/x-www-form-urlencoded
+
+$body
+EOF
+}
+
+sub http_patch {
+	my ($url, $body) = @_;
+	my $len = length($body);
+
+	http(<<EOF);
+PATCH $url HTTP/1.1
+Host: sp.example.com
+Connection: close
+Content-Length: $len
+
+$body
+EOF
+}
+
+sub get {
+    my ($uri, %extra) = @_;
+
+    my $auth_token = $extra{auth_token} || '';
+
+    http(<<EOF);
+GET $uri HTTP/1.0
+Host: sp.example.com
+Cookie: auth_token=$auth_token
+
+EOF
+}
+
+sub recode {
+	my $json = JSON::PP::decode_json(shift);
+	JSON::PP->new()->canonical()->encode($json);
+}
+
+sub getkv {
+	my ($uri) = @_;
+
+	get($uri) =~ /\x0d\x0a?\x0d\x0a?(.*)/ms;
+	recode($1);
```

Comment:
Style: add "return" to see what you are returning.

Reply:
Yes and no, style is such a thing that it's best to keep it the same across all of our perl tests.